tool/src/kotlin/mod.rs

```diff
@@ -594,23 +902,80 @@ if (returnVal == null) {{
             param_types_ffi.push(param_type_ffi);
             param_conversions.push(self.gen_kt_to_c_for_type(&param.ty, param_name.clone()));
         }
-        let write_return = matches!(&method.output, ReturnType::Infallible(SuccessType::Write));
+        let write_return = matches!(
+            &method.output,
+            ReturnType::Infallible(SuccessType::Write)
+                | ReturnType::Fallible(SuccessType::Write, _)
+                | ReturnType::Nullable(SuccessType::Write)
+        );
         if write_return {
             param_conversions.push("write".into());
         }
         let params = param_decls_kt.join(", ");
 
         let return_ty = self.gen_return_type_name(&method.output);
 
-        let declaration = format!(
-            "fun {}({}): {return_ty}",
-            self.formatter.fmt_method_name(method),
-            params
-        );
         let method_lifetimes_map = visitor.borrow_map();
         let return_expression = self
-            .gen_return(method, method_lifetimes_map, cleanups.as_ref())
-            .map(From::from);
+            .gen_return_conversion(method, method_lifetimes_map, cleanups.as_ref())
+            .into();
+
+        // this should only be called in the special method generation below
+        let non_option_type_name = |return_type: &ReturnType| match return_type {
+            ReturnType::Infallible(ok) | ReturnType::Nullable(ok) => {
+                self.gen_infallible_return_type_name(ok)
+            }
+            ReturnType::Fallible(_, _) => panic!(
+                "non_option_type_name should only be called for a return type that is optional"
+            ),
+        };
```

Comment:
@Manishearth this is the closure that was in the dedicated methods before. I could still drop the separate branch for the fallible method, as there's no need for the check.

Reply:
e.g.
```rust
        let infallible_type_name_pass_through = |return_type: &ReturnType| match return_type {
            ReturnType::Infallible(ok) | ReturnType::Nullable(ok) | ReturnType::Fallible(ok, _) => {
                self.gen_infallible_return_type_name(ok)
            }
        };

```

Reply:
I'd say up to you. I don't mind having duplicate checks as long as they're not overcomplicating the code.